File: bst.py
class BST:
    def __init__(self, root=None, c=0):
        self.c = c
        self.root = root

    def insert(self, node, number):
        if number > node.number and node.right is None:
            node.right = Node(number, node.height + 1)
            self.c += node.right.height
        elif number < node.number and node.left is None:
            node.left = Node(number, node.height + 1)
            self.c += node.left.height
        elif number > node.number and node.right is not None:
            self.insert(node.right, number)
        elif number < node.number and node.left is not None:
            self.insert(node.left, number)

    def get_c(self):
        return self.c

    def get_root(self):
        return self.root

class Node:
    def __init__(self, number=0, height=0):
        self.number = number
        self.height = height
        self.left = None
        self.right = None


if __name__ == "__main__":
    total = int(input())
    numbers = []
    for _ in range(total):
        numbers.append(int(input()))

    root = Node(numbers.pop(0))
    bst = BST(root)
    print(str(bst.get_c()))
    while numbers:
        bst.insert(bst.get_root(), numbers.pop(0))
        print(str(bst.get_c()))

File: bst.cpp
#include <iostream>
using namespace std;

struct Node{
    int number;
    int height;
    struct Node *left;
    struct Node *right;
};

struct Node *create_node(int digit, int height=0){
    struct Node *node = (struct Node*)malloc(sizeof(struct Node));
    node->number = digit;
    node->height = height;
    node->left = NULL;
    node->right = NULL;
    return node;
}

class BST{
    public:
        int c;
        Node *root;
        BST(Node *node){
            c = 0;
            root = node;
        }
        void insert_node(Node *node, int digit){
            if(digit > node->number && node->right == NULL){
                node->right = create_node(digit, (node->height) + 1);
                c += node->right->height;
            }
            else if(digit < node->number && node->left == NULL) {
                node->left = create_node(digit, (node->height) + 1);
                c += node->left->height;
            }
            else if(digit > node->number){
                insert_node(node->right, digit);
            }else{
                insert_node(node->left, digit);
            }
        }
};

int main(int argc, char *argv[]){
    int total;
    cin >> total;
    int numbers[total];

    for(int i = 0; i < total; i++){
        cin >> numbers[i];
    }

    cout << endl;

    BST bst = BST(create_node(numbers[0]));
    cout << bst.c << endl;
    for(int i = 1; i < total; i++){
        bst.insert_node(bst.root, numbers[i]);
        cout << bst.c << endl;
    }
}
